cmd/dagger/checks_frontend.go

```diff
@@ -0,0 +1,601 @@
+	"go.opentelemetry.io/otel/codes"
+	sdklog "go.opentelemetry.io/otel/sdk/log"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"golang.org/x/term"
+
+	"dagger.io/dagger"
+	"github.com/dagger/dagger/dagql/dagui"
+	"github.com/dagger/dagger/dagql/idtui"
+	"github.com/dagger/dagger/util/cleanups"
+)
+
+const (
+	tickInterval = 100 * time.Millisecond // Update frequency for TUI and animation
+)
+
+// Dots animation frames (fills up progressively)
+var dotsFrames = []rune{'⠀', '⡀', '⣀', '⣄', '⣤', '⣦', '⣶', '⣷', '⣿'}
+
+// checksFrontend is a live TUI frontend for displaying check execution
+type checksFrontend struct {
+	dagui.FrontendOpts
+	mu sync.Mutex
+
+	db      *ChecksDB
+	writer  io.Writer
+	profile termenv.Profile
+
+	// Check state tracking
+	checks         map[string]*checkState // key is check name
+	checkOrder     []string               // preserves check display order
+	animationFrame int                    // current frame of dots animation
+	lastEventTime  time.Time              // time of last span/log event (for activity detection)
+
+	// TUI state
+	program *tea.Program
+	done    bool
+	err     error
+
+	// Execution control
+	run      func(context.Context) (cleanups.CleanupF, error)
+	runCtx   context.Context
+	cleanup  cleanups.CleanupF
+	quitting bool
+
+	// Terminal state
+	width  int
+	height int
+}
+
+// checkState tracks the state of a single check
+type checkState struct {
+	name      string
+	status    checkStatus
+	span      *dagui.Span
+	checkInfo *CheckSpanInfo
+	logs      []LogWithContext
+	startTime time.Time
+	endTime   time.Time
+	passed    *bool
+}
+
+type checkStatus int
+
+const (
+	checkPending checkStatus = iota
+	checkRunning
+	checkPassed
+	checkFailed
+)
+
+func (s checkStatus) Emoji() string {
+	switch s {
+	case checkPending:
+		return "⚪"
+	case checkRunning:
+		return "⏳"
+	case checkPassed:
+		return "✅"
+	case checkFailed:
+		return "🔴"
+	default:
+		return "⚪"
+	}
+}
+
+// NewChecksFrontend creates a new live frontend for checks
+func NewChecksFrontend(w io.Writer, db *ChecksDB) *checksFrontend {
+	return &checksFrontend{
+		db:             db,
+		writer:         w,
+		profile:        termenv.ColorProfile(),
+		checks:         make(map[string]*checkState),
+		checkOrder:     []string{},
+		animationFrame: 0,
+	}
+}
+
+// Run implements the Frontend interface
+func (fe *checksFrontend) Run(ctx context.Context, opts dagui.FrontendOpts, f func(context.Context) (cleanups.CleanupF, error)) error {
+	fe.FrontendOpts = opts
+	fe.run = f
+	fe.runCtx = ctx
+
+	// Check if we have a TTY - if not, fallback to non-live mode
+	if !fe.isTTY() {
+		// Just run the function and return without TUI
+		cleanup, err := f(ctx)
+		if cleanup != nil {
+			defer cleanup()
+		}
+		return err
+	}
+
+	// Start the TUI with alternate screen (clears on exit)
+	fe.program = tea.NewProgram(fe, tea.WithAltScreen())
+
+	if _, err := fe.program.Run(); err != nil {
```

Comment:
Here, as you drop the returned `cleanup()` function, I think this will lead to a leak of the engine session, and skip the telemetry flush ? Right above, you seem to defer its cleanup (in the non TUI path)

Non blocking as everything is changing in this codepath, BUT, might be worth keeping in mind?